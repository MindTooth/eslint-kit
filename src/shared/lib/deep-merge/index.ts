/* eslint-disable @typescript-eslint/no-explicit-any */
import { DeepMerge, PathsOf } from './types'

function isObject(value: any): value is Record<string, any> {
  return typeof value === 'object' && value !== null
}

export enum Merge {
  Shallow,
  Deep,
  Override,
}

type AllPaths<A, B> = PathsOf<A> | PathsOf<B> | null

// eslint-disable-next-line sonarjs/cognitive-complexity
export function deepMerge<A extends any, B extends any>(
  a: A,
  b: B,
  strategy: (path: AllPaths<A, B>) => Merge = () => Merge.Deep,
  basePath: string | null = null,
  root = true,
  shallow = false
): DeepMerge<A, B> {
  if (Array.isArray(a) && Array.isArray(b)) {
    return [...a, ...b] as DeepMerge<A, B>
  }

  if (isObject(a) && isObject(b)) {
    if (shallow) {
      return { ...a, ...b } as DeepMerge<A, B>
    }

    if (root) {
      if (strategy(null) === Merge.Shallow) {
        return deepMerge(a, b, strategy as any, null, false, true) as DeepMerge<
          A,
          B
        >
      }

      if (strategy(null) === Merge.Override) {
        return b as DeepMerge<A, B>
      }
    }

    const keys = [...Object.keys(a), ...Object.keys(b)]
    const final: any = {}

    for (const key of keys) {
      const path = basePath ? [basePath, key].join('.') : key

      if (strategy(path as any) === Merge.Shallow) {
        final[key] = deepMerge(
          a[key],
          b[key],
          strategy as any,
          path,
          false,
          true
        )
      } else if (strategy(path as any) === Merge.Override) {
        final[key] = b[key]
      } else {
        final[key] = deepMerge(a[key], b[key], strategy as any, path, false)
      }
    }

    return final
  }

  if (typeof b !== 'undefined') {
    return b as DeepMerge<A, B>
  }

  return a as DeepMerge<A, B>
}
