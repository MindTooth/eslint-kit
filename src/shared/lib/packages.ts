import { parse } from 'semver'

type Dependencies = Record<string, string>

export interface PackageJson {
  version: string
  dependencies?: Dependencies
  devDependencies?: Dependencies
}

function mergeDependencies(packageJson: PackageJson) {
  const { dependencies = {}, devDependencies = {} } = packageJson
  return { ...dependencies, ...devDependencies }
}

export function getPackageSemver(name: string, packageJson: PackageJson) {
  const allDependencies = mergeDependencies(packageJson)
  const version = allDependencies[name]
  if (!version) return null
  const parsed = parse(version)
  if (!parsed) return null
  return parsed
}
