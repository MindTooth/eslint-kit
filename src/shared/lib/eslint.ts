import { Linter } from 'eslint'
import { deepMerge, Strategy } from './deep-merge'

export function mergeConfigs(configs: Linter.Config[]): Linter.Config {
  return configs.reduce((final, current) => {
    return deepMerge(final, current, (path) => {
      if (path === 'rules') return Strategy.Shallow
      return Strategy.Deep
    })
  })
}
