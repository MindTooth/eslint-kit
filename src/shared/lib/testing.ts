import fs from 'fs'
import path from 'path'
import { ESLint, Linter } from 'eslint'

export interface Options {
  config: unknown
  basePath: string
  files: string[]
  extension: 'js' | 'ts' | 'tsx'
}

export async function testConfig({
  config,
  basePath,
  files,
  extension,
}: Options) {
  const cli = new ESLint({
    baseConfig: config as Linter.Config<Linter.RulesRecord>,
    useEslintrc: false,
    cwd: process.cwd(),
    ignore: false,
  })

  for (const file of files) {
    const filePath = path.resolve(basePath, `./tests/${file}.${extension}`)
    const code = fs.readFileSync(filePath).toString()
    const isTS = ['ts', 'tsx'].includes(extension)
    const result = await cli.lintText(code, {
      filePath: isTS ? filePath : undefined,
    })
    expect(result).toMatchSnapshot(`${file}`)
  }
}
