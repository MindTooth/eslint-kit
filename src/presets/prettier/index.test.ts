import { mergeConfigs } from '../../shared/lib/eslint'
import { testConfig } from '../../shared/lib/testing'
import { presetBase } from '../base'
import { presetPrettier } from './index'

describe('[Presets] Prettier', () => {
  it('should match snapshots', async () => {
    await testConfig({
      config: mergeConfigs([presetBase, presetPrettier]),
      basePath: __dirname,
      files: ['basic-clear', 'basic-warn'],
      extension: 'js',
    })
  })
})
