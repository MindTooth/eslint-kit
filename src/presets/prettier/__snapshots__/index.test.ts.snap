// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Presets] Prettier should match snapshots: basic-clear 1`] = `
Array [
  Object {
    "errorCount": 0,
    "fatalErrorCount": 0,
    "filePath": "<PROJECT_ROOT>/src/presets/prettier/tests/basic-clear.js",
    "fixableErrorCount": 0,
    "fixableWarningCount": 0,
    "messages": Array [],
    "suppressedMessages": Array [],
    "usedDeprecatedRules": Array [],
    "warningCount": 0,
  },
]
`;

exports[`[Presets] Prettier should match snapshots: basic-warn 1`] = `
Array [
  Object {
    "errorCount": 0,
    "fatalErrorCount": 0,
    "filePath": "<PROJECT_ROOT>/src/presets/prettier/tests/basic-warn.js",
    "fixableErrorCount": 0,
    "fixableWarningCount": 1,
    "messages": Array [
      Object {
        "column": 24,
        "endColumn": 25,
        "endLine": 1,
        "fix": Object {
          "range": Array [
            23,
            24,
          ],
          "text": "
",
        },
        "line": 1,
        "message": "Replace \`;\` with \`‚èé\`",
        "messageId": "replace",
        "nodeType": null,
        "ruleId": "prettier/prettier",
        "severity": 1,
      },
    ],
    "source": "export const meow = 123;",
    "suppressedMessages": Array [],
    "usedDeprecatedRules": Array [],
    "warningCount": 1,
  },
]
`;
